File: backend/app/schemas/journey.py
import uuid
from datetime import datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field, ConfigDict

class JourneyActionDefinition(BaseModel):
    action_id: str
    label: str
    target_stage: str
    target_macro_stage: str
    required_role: str = "sales"
    requires_evidence: bool = False
    evidence_prompt: Optional[str] = None
    description: str

class BlockedActionDefinition(BaseModel):
    action_id: str
    label: str
    target_stage: str
    blocked_reasons: List[str]
    prerequisites: List[str]

class StageEventResponse(BaseModel):
    id: uuid.UUID
    entity_type: str
    entity_id: uuid.UUID
    macro_stage: str
    previous_stage: str
    new_stage: str
    action: str
    actor: str
    actor_role: str
    reason_code: str
    notes: Optional[str] = None
    evidence_references: Dict[str, Any] = Field(default_factory=dict)
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)

class JourneyStateResponse(BaseModel):
    entity_id: uuid.UUID
    entity_type: str
    current_stage: str
    macro_stage: str
    stage_title: str
    owner_question: str
    available_actions: List[JourneyActionDefinition]
    blocked_actions: List[BlockedActionDefinition]
    history: List[StageEventResponse]

class JourneyTransitionRequest(BaseModel):
    action_id: str
    actor: str = "Johann Butler"
    actor_role: str = "owner"
    reason_code: str = "owner_decision"
    notes: Optional[str] = None
    evidence_references: Dict[str, Any] = Field(default_factory=dict)
    idempotency_key: Optional[str] = None

class JourneyTransitionResponse(BaseModel):
    success: bool
    entity_id: uuid.UUID
    previous_stage: str
    new_stage: str
    macro_stage: str
    event_id: uuid.UUID
    message: str
    available_actions: List[JourneyActionDefinition]

File: frontend/src/api/journey.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { fetchApi } from './client';

export interface JourneyActionDefinition {
  action_id: string;
  label: string;
  target_stage: string;
  target_macro_stage: string;
  required_role: string;
  requires_evidence: boolean;
  evidence_prompt?: string;
  description: string;
}

export interface BlockedActionDefinition {
  action_id: string;
  label: string;
  target_stage: string;
  blocked_reasons: string[];
  prerequisites: string[];
}

export interface StageEvent {
  id: string;
  entity_type: string;
  entity_id: string;
  macro_stage: string;
  previous_stage: string;
  new_stage: string;
  action: string;
  actor: string;
  actor_role: string;
  reason_code: string;
  notes?: string;
  evidence_references: Record<string, any>;
  created_at: string;
}

export interface JourneyState {
  entity_id: string;
  entity_type: string;
  current_stage: string;
  macro_stage: string;
  stage_title: string;
  owner_question: string;
  available_actions: JourneyActionDefinition[];
  blocked_actions: BlockedActionDefinition[];
  history: StageEvent[];
}

export interface TransitionRequest {
  action_id: string;
  actor?: string;
  actor_role?: string;
  reason_code?: string;
  notes?: string;
  evidence_references?: Record<string, any>;
  idempotency_key?: string;
}

export interface TransitionResponse {
  success: boolean;
  entity_id: string;
  previous_stage: string;
  new_stage: string;
  macro_stage: string;
  event_id: string;
  message: string;
  available_actions: JourneyActionDefinition[];
}

export function useJourneyState(oppId?: string) {
  return useQuery<JourneyState>({
    queryKey: ['journeyState', oppId],
    queryFn: () => fetchApi<JourneyState>(`/api/v1/journey/opportunities/${oppId}/state`),
    enabled: !!oppId,
  });
}

export function useJourneyHistory(oppId?: string) {
  return useQuery<StageEvent[]>({
    queryKey: ['journeyHistory', oppId],
    queryFn: () => fetchApi<StageEvent[]>(`/api/v1/journey/opportunities/${oppId}/history`),
    enabled: !!oppId,
  });
}

export function useExecuteTransition() {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: ({ oppId, req }: { oppId: string; req: TransitionRequest }) =>
      fetchApi<TransitionResponse>(`/api/v1/journey/opportunities/${oppId}/transition`, {
        method: 'POST',
        body: JSON.stringify(req),
      }),
    onSuccess: (_, { oppId }) => {
      queryClient.invalidateQueries({ queryKey: ['deals'] });
      queryClient.invalidateQueries({ queryKey: ['journeyState', oppId] });
      queryClient.invalidateQueries({ queryKey: ['journeyHistory', oppId] });
      queryClient.invalidateQueries({ queryKey: ['todayCockpit'] });
      queryClient.invalidateQueries({ queryKey: ['pipelineSummary'] });
    },
  });
}
